File: backend-node/src/services/comandaService.js
const { v4: uuidv4 } = require('uuid');
const { EventType } = require('../models');
const { nowIso } = require('../utils');
const config = require('../config');

// Validates the expected code format F001, F002, ..., F999, F1000, etc.
const CODE_REGEX = /^F(\d+)$/;

class MaxComandasReachedError extends Error {}
class DuplicateCodeError extends Error {}

function getNextCode(db) {
  const row = db.prepare('SELECT code FROM comandas ORDER BY created_at DESC LIMIT 1').get();
  if (!row) return 'F001';
  const match = CODE_REGEX.exec(row.code);
  if (!match) return 'F001';
  const number = parseInt(match[1], 10);
  return `F${String(number + 1).padStart(3, '0')}`;
}

function createComanda(db, holderName, initialBalance) {
  const trimmedName = holderName ? String(holderName).trim() : '';
  if (!trimmedName) throw new Error('holder_name is required');
  if (!Number.isInteger(initialBalance) || initialBalance < 0) {
    throw new Error('initial_balance must be a non-negative integer');
  }

  const count = db.prepare('SELECT COUNT(*) as c FROM comandas').get().c;
  if (count >= config.maxComandas) {
    throw new MaxComandasReachedError(
      `Maximum number of comandas (${config.maxComandas}) reached`
    );
  }

  const comandaId = uuidv4();
  const code = getNextCode(db);
  const createdAt = nowIso();
  const eventId = uuidv4();

  const insertComanda = db.prepare(
    'INSERT INTO comandas (id, code, holder_name, created_at) VALUES (?, ?, ?, ?)'
  );
  const insertEvent = db.prepare(
    'INSERT INTO events (id, type, comanda_id, amount, note, timestamp) VALUES (?, ?, ?, ?, ?, ?)'
  );

  const transaction = db.transaction(() => {
    insertComanda.run(comandaId, code, trimmedName, createdAt);
    if (initialBalance > 0) {
      insertEvent.run(eventId, EventType.credit, comandaId, initialBalance, 'Saldo inicial', createdAt);
    }
  });

  try {
    transaction();
  } catch (err) {
    if (err.message && err.message.includes('UNIQUE constraint failed: comandas.code')) {
      throw new DuplicateCodeError('Concurrent comanda creation conflict. Please retry.');
    }
    throw err;
  }

  return { comanda: { id: comandaId, code, holder_name: trimmedName, created_at: createdAt }, eventId };
}

function getComandaByCode(db, code) {
  return db.prepare('SELECT * FROM comandas WHERE code = ?').get(code) || null;
}

function getBalance(db, comandaId) {
  const row = db.prepare('SELECT balance FROM balance_view WHERE comanda_id = ?').get(comandaId);
  return row && row.balance !== null ? row.balance : 0;
}

function getComandaEvents(db, comandaId) {
  return db.prepare('SELECT * FROM events WHERE comanda_id = ? ORDER BY timestamp ASC').all(comandaId);
}

module.exports = {
  getNextCode,
  createComanda,
  getComandaByCode,
  getBalance,
  getComandaEvents,
  MaxComandasReachedError,
  DuplicateCodeError,
};

File: backend-python/app/services/comanda_service.py
import uuid
import re
import sqlite3
from typing import Optional
from app.models import Comanda, Event, EventType
from app.utils import now_iso
from app.config import settings

CODE_REGEX = re.compile(r'^F(\d+)$')


class MaxComandasReachedError(Exception):
    pass


class DuplicateCodeError(Exception):
    pass


def get_next_code(conn) -> str:
    """Retorna o próximo código disponível (ex: F001, F002)"""
    cursor = conn.cursor()
    cursor.execute("SELECT code FROM comandas ORDER BY created_at DESC LIMIT 1")
    row = cursor.fetchone()

    if not row:
        return "F001"

    last_code = row["code"]
    match = CODE_REGEX.match(last_code)
    if not match:
        return "F001"
    number = int(match.group(1))
    return f"F{number + 1:03d}"

def create_comanda(conn, holder_name: str, initial_balance: int) -> tuple[Comanda, str]:
    """Cria uma comanda gerando um código automático e emite saldo inicial."""
    trimmed_name = holder_name.strip() if holder_name else ''
    if not trimmed_name:
        raise ValueError('holder_name is required')
    if not isinstance(initial_balance, int) or initial_balance < 0:
        raise ValueError('initial_balance must be a non-negative integer')

    cursor = conn.cursor()
    cursor.execute("SELECT COUNT(*) as c FROM comandas")
    count = cursor.fetchone()["c"]
    if count >= settings.max_comandas:
        raise MaxComandasReachedError(
            f"Maximum number of comandas ({settings.max_comandas}) reached"
        )

    comanda_id = str(uuid.uuid4())
    code = get_next_code(conn)
    created_at = now_iso()

    try:
        cursor.execute(
            "INSERT INTO comandas (id, code, holder_name, created_at) VALUES (?, ?, ?, ?)",
            (comanda_id, code, trimmed_name, created_at)
        )

        event_id = str(uuid.uuid4())
        if initial_balance > 0:
            cursor.execute(
                "INSERT INTO events (id, type, comanda_id, amount, note, timestamp) VALUES (?, ?, ?, ?, ?, ?)",
                (event_id, EventType.credit.value, comanda_id, initial_balance, "Saldo inicial", created_at)
            )

        conn.commit()
    except sqlite3.IntegrityError as err:
        if 'UNIQUE constraint failed: comandas.code' in str(err):
            raise DuplicateCodeError('Concurrent comanda creation conflict. Please retry.')
        raise

    comanda = Comanda(id=comanda_id, code=code, holder_name=trimmed_name, created_at=created_at)
    return comanda, event_id

def get_comanda_by_code(conn, code: str) -> Optional[Comanda]:
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM comandas WHERE code = ?", (code,))
    row = cursor.fetchone()
    if row:
        return Comanda(**dict(row))
    return None

def get_balance(conn, comanda_id: str) -> int:
    """Consulta a view 'balance_view' que processa on-the-fly a soma dos eventos."""
    cursor = conn.cursor()
    cursor.execute("SELECT balance FROM balance_view WHERE comanda_id = ?", (comanda_id,))
    row = cursor.fetchone()
    if row and row["balance"] is not None:
        return row["balance"]
    return 0

def get_comanda_events(conn, comanda_id: str) -> list[Event]:
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM events WHERE comanda_id = ? ORDER BY timestamp ASC", (comanda_id,))
    events = []
    for row in cursor.fetchall():
        events.append(Event(**dict(row)))
    return events
